poll/poll.py

```diff
@@ -0,0 +1,849 @@
+# along with this program in a file in the toplevel directory called
+# "AGPLv3".  If not, see <http://www.gnu.org/licenses/>.
+#
+
+from collections import OrderedDict
+
+from django.template import Template, Context
+from markdown import markdown
+
+import pkg_resources
+
+from xblock.core import XBlock
+from xblock.fields import Scope, String, Dict, List, Boolean, Integer
+from xblock.fragment import Fragment
+from xblockutils.publish_event import PublishEventMixin
+from xblockutils.resources import ResourceLoader
+
+HAS_EDX_ACCESS = False
+try:
+    # pylint: disable=import-error
+    from django.conf import settings
+    from courseware.access import has_access
+    from api_manager.models import GroupProfile
+    HAS_EDX_ACCESS = True
+except ImportError:
+    pass
+
+
+class ResourceMixin(object):
+    loader = ResourceLoader(__name__)
+
+    @staticmethod
+    def resource_string(path):
+        """Handy helper for getting resources from our kit."""
+        data = pkg_resources.resource_string(__name__, path)
+        return data.decode("utf8")
+
+    def create_fragment(self, context, template, css, js, js_init):
+        html = Template(
+            self.resource_string(template)).render(Context(context))
+        frag = Fragment(html)
+        frag.add_javascript_url(
+            self.runtime.local_resource_url(
+                self, 'public/js/vendor/handlebars.js'))
+        frag.add_css(self.resource_string(css))
+        frag.add_javascript(self.resource_string(js))
+        frag.initialize_js(js_init)
+        return frag
+
+
+class PollBase(XBlock, ResourceMixin, PublishEventMixin):
+    """
+    Base class for Poll-like XBlocks.
+    """
+    event_namespace = 'xblock.pollbase'
+    private_results = Boolean(default=False, help="Whether or not to display results to the user.")
+    max_submissions = Integer(default=1, help="The maximum number of times a user may send a submission.")
+    submissions_count = Integer(
+        default=0, help="Number of times the user has sent a submission.", scope=Scope.user_state
+    )
+    feedback = String(default='', help="Text to display after the user votes.")
+
+    def send_vote_event(self, choice_data):
+        # Let the LMS know the user has answered the poll.
+        self.runtime.publish(self, 'progress', {})
+        self.runtime.publish(self, 'grade', {
+            'value': 1,
+            'max_value': 1,
+            }
+        )
+        # The SDK doesn't set url_name.
+        event_dict = {'url_name': getattr(self, 'url_name', '')}
+        event_dict.update(choice_data)
+        self.publish_event_from_dict(
+            self.event_namespace + '.submitted',
+            event_dict,
+        )
+
+    @staticmethod
+    def any_image(field):
+        """
+        Find out if any answer has an image, since it affects layout.
+        """
+        return any(value['img'] for value in dict(field).values())
```

Comment:
A small point: there are many places where the field is converted from lists to dicts, etc.  This could be avoided here with:
`any(value['img'] for key, value in field)`